js/ui/status/keyboard.js

```diff
@@ -480,6 +484,51 @@ const InputSourceManager = new Lang.Class({
             this._updateMruSettings();
     },
 
+    get passwordModeEnabled() {
+        return this._passwordModeEnabled;
+    },
+
+    set passwordModeEnabled(enable) {
+        this._passwordModeEnabled = enable;
+        this._syncPasswordMode();
+    },
+
+    _syncPasswordMode: function() {
+        let source;
+
+        if (this._passwordModeEnabled) {
+            // do not switch the keyboard layout if the current one is a latin one
+            if (this._keyboardManager.isLatinLayout(this._currentSource.id))
+                return;
+
+            // look for latin keyboard layout to switch to
+            for (let i = 0; i < this._mruSources.length; ++i) {
+                let src = this._mruSources[i];
+                if (this._keyboardManager.isLatinLayout(src.id)) {
+                    source = src;
+                    break;
+                }
+            }
+            if (!source) {
+                // fallback to an 'en' keyboard layout
+                this._needsFallbackSource = true;
+                this._inputSourcesChanged();
+                source = this._fallbackInputSource;
+            }
+            this._lastSource = this._currentSource;
+            this._activateInputSource(source, false);
+        } else {
+            if (!this._lastSource)
+                return;
+            source = this._lastSource;
+            this._activateInputSource(source, false);
+            this._lastSource = null;
+            this._needsFallbackSource = false;
+            this._fallbackInputSource = null;
```

Comment:
Since `this._inputSourcesChanged()` sets `this._fallbackInputSource` when `this._needsFallbackSource == true`, it should also unset it when `this._needsFallbackSource == false`. On a related note, you could also pass directly `needsFallbackSoruce` to `this._inputSourcesChanged()` instead of saving it as a private member. Or alternatively, if `this._inputSourcesChanged()` is called from too many places to make that practical, you can move the body of that function to another method that takes the parameter, and call the new method from both places.

Reply:
Indeed, Isolved this by adding a default param to this._inputSourcesChanged(), so I only need to pass the flag from _syncPasswordMode.